target/linux/rtl838x/files-5.4/drivers/net/dsa/rtl838x_sw.c

```diff
@@ -0,0 +1,2172 @@
+		printk("PHY-read from SDS: port %d reg: %x res: %x\n", phy_addr, phy_reg, val);
+		return val;
+	}
+
+	read_phy(phy_addr, 0, phy_reg, &val);
+//	printk("PHY-read: port %d reg: %x res: %x\n", phy_addr, phy_reg, val);
+	return val;
+}
+
+int dsa_phy_write(struct dsa_switch *ds, int phy_addr, int phy_reg, u16 val)
+{
+	u32 offset = 0;
+	struct rtl838x_switch_priv *priv = ds->priv;
+	
+	if (phy_addr >= 24 && phy_addr <= 27 
+	     && priv->ports[24].phy == PHY_RTL838X_SDS) {
+//		printk("PHY_write to SDS, port %d\n", phy_addr);
+		if (phy_addr == 26)
+			offset = 0x100;
+		sw_w32(val, MAPLE_SDS4_FIB_REG0r + offset + (phy_reg << 2));
+		return 0;
+	}
+	return write_phy(phy_addr, 0, phy_reg, val);
+}
+
+static int rtl838x_mdio_read(struct mii_bus *bus, int addr, int regnum)
+{
+	int ret;
+	
+	struct rtl838x_switch_priv *priv = bus->priv;
+//	printk("switch rtl838x_mdio_read %d %d\n", addr, regnum);
+	ret = dsa_phy_read(priv->ds, addr, regnum);
+//	printk("switch rtl838x_mdio_read %d\n", ret);
+	return ret;
+}
+
+static int rtl838x_mdio_write(struct mii_bus *bus, int addr, int regnum,
+				 u16 val)
+{
+	struct rtl838x_switch_priv *priv = bus->priv;
+	return dsa_phy_write(priv->ds, addr, regnum, val);
+}
+
+static void rtl838x_enable_phy_polling(struct rtl838x_switch_priv *priv)
+{
+	int i;
+	u64 v = 0;
+
+	msleep(1000);
+	/* Enable all ports with a PHY, including the SFP-ports */
+	for (i = 0; i < priv->cpu_port; i++) {
+		if (priv->ports[i].phy) v |= 1 << i;
+	}
+	printk("rtl83xx_enable_phy_polling: %16llx\n", v);
+	priv->r->set_port_reg(v, priv->r->smi_poll_ctrl);
+
+	/* PHY update complete */
+	if (priv->family_id == RTL8390_FAMILY_ID)
+		sw_w32_mask(1 << 7, 0, RTL839X_SMI_GLB_CTRL);
+	else
+		sw_w32_mask(0, 0x8000, RTL838X_SMI_GLB_CTRL);
+}
+
+void rtl839x_print_matrix(void)
+{
+	volatile u64 *ptr = RTL839X_PORT_ISO_CTRL(0);
+	int i;
+
+	for (i = 0; i < 52; i += 4)
+		printk("> %16llx %16llx %16llx %16llx\n",
+			ptr[i + 0], ptr[i+ 1], ptr[i + 2], ptr[i + 3]);
+	printk("CPU_PORT> %16llx\n", ptr[52]);
+}
+
+void rtl838x_print_matrix(void)
+{
+	unsigned volatile int *ptr = RTL838X_PORT_ISO_CTRL(0);
+	int i;
+
+	if (soc_info.family == RTL8390_FAMILY_ID)
+		return rtl839x_print_matrix();
+
+	for (i = 0; i < 28; i += 8)
+	printk("> %8x %8x %8x %8x %8x %8x %8x %8x\n", 
+		ptr[i + 0], ptr[i + 1], ptr[i + 2], ptr[i + 3], ptr[i + 4], ptr[i + 5],
+		ptr[i + 6], ptr[i + 7]);
+	printk("CPU_PORT> %8x\n", ptr[28]);
+}
+
+static void rtl838x_init_stats(struct rtl838x_switch_priv *priv)
+{
+	mutex_lock(&priv->reg_mutex);
+
+	/* Enable statistics module: all counters plus debug.
+	   On RTL839x all counters are enabled by default */
+	if (priv->family_id == RTL8380_FAMILY_ID)
+		sw_w32_mask(0, 3, RTL838X_STAT_CTRL);
+
+	/* Reset statistics counters */
+	sw_w32_mask(0, 1, priv->r->stat_rst);
+
+	mutex_unlock(&priv->reg_mutex);
+}
+
+static int rtl838x_setup(struct dsa_switch *ds)
+{
+	int i;
+	u64 port_bitmap = 0;
+	struct rtl838x_switch_priv *priv = ds->priv;
```

Comment:
Does this switch support setting up VLANs when VLAN filtering is disabled?
If yes, add `ds->configure_vlan_while_not_filtering = true;` here

Reply:
Yes, it does. I added the flag.